joblib/parallel.py

```diff
@@ -367,23 +382,170 @@ class BatchCompletionCallBack(object):
     processing duration and to schedule the next batch of tasks to be
     processed.
 
+    It is assumed that this callback will always be triggered by the backend
+    right after the end of a task, in case of success and in case of failure.
+    """
+
+    """
+    ##########################################################################
+    ##################### METHODS CALLED BY THE MAIN THREAD ##################
+    ##########################################################################
     """
     def __init__(self, dispatch_timestamp, batch_size, parallel):
         self.dispatch_timestamp = dispatch_timestamp
         self.batch_size = batch_size
         self.parallel = parallel
+        self.parallel_call_id = parallel._call_id
+
+        # Internals to keep track of the status and outcome of the task.
+
+        # Right after a job has been scheduled with this callback, the
+        # reference to the job object can be registered with `register_job`
+        self.job = None
+
+        # The latest known status for the job, can be TASK_PENDING, TASK_DONE,
+        # or TASK_ERROR
+        self.status = TASK_PENDING
+        if not parallel._backend.supports_retrieve_callback:
+            self.status = None
+
+    def register_job(self, job):
+        """Register the object returned by `apply_async`."""
+        self.job = job
+
+    def get_result(self, timeout):
+        backend = self.parallel._backend
+
+        if backend.supports_retrieve_callback:
+            # The result has already been retrieved by the callback thread, and
+            # is stored internally. It's just waiting to be returned.
+            return self._return_or_raise()
+
+        # For other backends, the main thread still need to run the retrieval
+        # step
+        try:
+            if backend.supports_timeout:
+                result = self.job.get(timeout=timeout)
+            else:
+                result = self.job.get()
+            outcome = dict(result=result, status=TASK_DONE)
+        except BaseException as e:
+            outcome = dict(result=e, status=TASK_ERROR)
+        self._register_outcome(outcome)
 
+        return self._return_or_raise()
+
+    def _return_or_raise(self):
+        try:
+            if self.status == TASK_ERROR:
+                raise self._result
+            return self._result
+        finally:
+            del self._result
+
+    def get_status(self, timeout):
+        if timeout is None or self.status != TASK_PENDING:
+            return self.status
+
+        # The computation are running and the status is pending.
+        # Check that we did not wait for this jobs more than `timeout`.
+        now = time.time()
+        if not hasattr(self, "_completion_timeout_counter"):
+            self._completion_timeout_counter = now
```

Comment:
It's true that it's not exactly the same logic when the backend supports retrieving in the callback thread. By I think that in effect, both implementations will result in almost-identical behaviors. In both cases, the timeout counter is triggered when the generator is iterated on. The only difference I can see is that the timeout step here is at most `0.01` seconds (corresponding to the `sleep` in `Parallel._retrieve`) while for backends that do not support the retrieve callback, the backend can be arbitrarily accurate regarding the timestep.

Reply:
Maybe it will be clearer to you after I push a docstring to `get_results` + reading `Parallel._retrieve`. We can make it even clearer checking in `get_result` the assumptions that are stated in the (soon to be pushed) docstring and raise `AssertionError` ?

Reply:
I propose we do a separate issue on this an just improve doc on the timeout behavior.